File: Hearthstone-Script/src/main/java/club/xiaojiawei/controller/javafx/StrategySettingsController.java
package club.xiaojiawei.controller.javafx;

import club.xiaojiawei.controls.NotificationManager;
import club.xiaojiawei.controls.NumberField;
import club.xiaojiawei.enums.ConfigEnum;
import club.xiaojiawei.enums.WindowEnum;
import club.xiaojiawei.utils.ConfigUtil;
import club.xiaojiawei.utils.WindowUtil;
import javafx.beans.value.ChangeListener;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.Scene;
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.VBox;

import java.net.URL;
import java.util.ResourceBundle;

/**
 *
 * @author 肖嘉威
 * @date 2023/9/10 15:07
 */
public class StrategySettingsController implements Initializable {

    @FXML
    private AnchorPane rootPane;
    @FXML
    private NumberField actionIntervalField;
    @FXML
    private NotificationManager<Object> notificationManager;
    @FXML
    private NumberField mouseMoveIntervalField;
    @FXML
    private VBox mainVBox;

    private ChangeListener<Scene> sceneListener;

    @Override
    public void initialize(URL url, ResourceBundle resourceBundle) {
        initStructure();
        initValue();
        listen();
    }

    private void initStructure(){
        actionIntervalField.setMinValue("1");
        actionIntervalField.setPromptText("默认：" + ConfigEnum.MOUSE_ACTION_INTERVAL.getDefaultValue());
        mouseMoveIntervalField.setMinValue("1");
        mouseMoveIntervalField.setPromptText("默认：" + ConfigEnum.MOUSE_MOVE_INTERVAL.getDefaultValue());
    }

    private void initValue(){
        actionIntervalField.setText(ConfigUtil.INSTANCE.getString(ConfigEnum.MOUSE_ACTION_INTERVAL));
        mouseMoveIntervalField.setText(ConfigUtil.INSTANCE.getString(ConfigEnum.MOUSE_MOVE_INTERVAL));
    }

    private void listen(){
        sceneListener = (observableValue, scene, t1) -> {
            mainVBox.prefWidthProperty().bind(t1.widthProperty());
            mainVBox.sceneProperty().removeListener(sceneListener);
        };
        mainVBox.sceneProperty().addListener(sceneListener);
    }

    @FXML
    protected void apply(ActionEvent actionEvent) {
        if (saveProperties()){
            notificationManager.showSuccess("应用成功", 2);
        }else {
            notificationManager.showWarn("应用失败", "值不合法",  2);
        }
    }

    @FXML
    protected void save(ActionEvent actionEvent) {
        if (saveProperties()){
            WindowUtil.INSTANCE.hideStage(WindowEnum.SETTINGS);
        }else {
            notificationManager.showWarn("保存失败", "值不合法",  2);
        }
    }

    private boolean saveProperties(){
        String actionInterval = actionIntervalField.getText();
        if (actionInterval == null || actionInterval.isBlank()){
            return false;
        }
        ConfigUtil.INSTANCE.putString(ConfigEnum.MOUSE_ACTION_INTERVAL, actionInterval, true);
        String mouseMoveIntervalFieldText = mouseMoveIntervalField.getText();
        if (mouseMoveIntervalFieldText == null || mouseMoveIntervalFieldText.isBlank()){
            return false;
        }
        ConfigUtil.INSTANCE.putString(ConfigEnum.MOUSE_MOVE_INTERVAL, mouseMoveIntervalFieldText, true);
        return true;
    }
}

File: Hearthstone-Script/src/main/java/club/xiaojiawei/controller/javafx/StrategySettingsController.kt
package club.xiaojiawei.controller.javafx

import club.xiaojiawei.utils.PropertiesUtil
import jakarta.annotation.Resource
import javafx.beans.value.ChangeListener
import javafx.event.ActionEvent
import javafx.fxml.Initializable
import org.springframework.stereotype.Component
import java.net.URL
import java.util.Properties

/**
 *
 * @author 肖嘉威
 * @date 2023/9/10 15:07
 */
class StrategySettingsController : Initializable {

    @Resource
    private val scriptConfiguration: Properties? = null

    @Resource
    private val propertiesUtil: PropertiesUtil? = null

    private var sceneListener: ChangeListener<Scene?>? = null

    @FXML
    private val rootPane: AnchorPane? = null

    @FXML
    private val actionIntervalField: NumberField? = null

    @FXML
    private val notificationManager: NotificationManager<*>? = null

    @FXML
    private val mouseMoveIntervalField: NumberField? = null

    @FXML
    private val mainVBox: VBox? = null

    override fun initialize(url: URL?, resourceBundle: ResourceBundle?) {
        initStructure()
        initValue()
        listen()
    }

    private fun initStructure() {
        actionIntervalField.setMinValue("1")
        actionIntervalField.setPromptText("默认：" + ConfigEnum.MOUSE_ACTION_INTERVAL.defaultValue)
        mouseMoveIntervalField.setMinValue("1")
        mouseMoveIntervalField.setPromptText("默认：" + ConfigEnum.MOUSE_MOVE_INTERVAL.defaultValue)
    }

    private fun initValue() {
        actionIntervalField.setText(
            scriptConfiguration!!.getProperty(
                ConfigEnum.MOUSE_ACTION_INTERVAL.getKey(),
                ConfigEnum.MOUSE_ACTION_INTERVAL.defaultValue
            )
        )
        mouseMoveIntervalField.setText(
            scriptConfiguration.getProperty(
                ConfigEnum.MOUSE_MOVE_INTERVAL.getKey(),
                ConfigEnum.MOUSE_MOVE_INTERVAL.defaultValue
            )
        )
    }

    private fun listen() {
        sceneListener = ChangeListener { observableValue: ObservableValue<out Scene?>?, scene: Scene?, t1: Scene? ->
            mainVBox.prefWidthProperty().bind(t1.widthProperty())
            mainVBox.sceneProperty().removeListener(sceneListener)
        }
        mainVBox.sceneProperty().addListener(sceneListener)
    }

    @FXML
    protected fun apply(actionEvent: ActionEvent?) {
        if (saveProperties()) {
            notificationManager.showSuccess("应用成功", 2)
        } else {
            notificationManager.showWarn("应用失败", "值不合法", 2)
        }
    }

    @FXML
    protected fun save(actionEvent: ActionEvent?) {
        if (saveProperties()) {
            WindowUtil.hideStage(WindowEnum.SETTINGS)
        } else {
            notificationManager.showWarn("保存失败", "值不合法", 2)
        }
    }

    private fun saveProperties(): Boolean {
        val actionInterval: String? = actionIntervalField.getText()
        if (actionInterval == null || actionInterval.isBlank()) {
            return false
        }
        scriptConfiguration!!.setProperty(ConfigEnum.MOUSE_ACTION_INTERVAL.getKey(), actionInterval)
        val mouseMoveIntervalFieldText: String? = mouseMoveIntervalField.getText()
        if (mouseMoveIntervalFieldText == null || mouseMoveIntervalFieldText.isBlank()) {
            return false
        }
        scriptConfiguration.setProperty(ConfigEnum.MOUSE_MOVE_INTERVAL.getKey(), mouseMoveIntervalFieldText)
        propertiesUtil.storeScriptProperties()
        return true
    }
}
